dashdblocal_notebooks/livy-notebooks/IoT_demo_Scala_new.ipynb

```diff
@@ -0,0 +1,1302 @@
+    "            );\n",
+    "    processMessages(messages)\n",
+    "    \n",
+    "    try {\n",
+    "        ssc.start()\n",
+    "        timeoutMillisecs match {\n",
+    "            case Some(t) => ssc.awaitTerminationOrTimeout(t)\n",
+    "            case _ => ssc.awaitTermination()\n",
+    "        }\n",
+    "    } finally {\n",
+    "        ssc.stop(false, true)\n",
+    "    }\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "When running interactively in the notebook we want all streaming tests to run only for 10 seconds and stop automatically. Later, when we deploy the notebook as a stand-alone Spark application, we want the processing to continue until the application is stopped externally. Hence we tag the cell NOT-FOR-APP."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "application/vnd.jupyter.widget-view+json": {
+       "model_id": "",
+       "version_major": 2,
+       "version_minor": 0
+      },
+      "text/plain": [
+       "FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "timeoutMillisecs: Option[Long] = Some(10000)\n"
+     ]
+    }
+   ],
+   "source": [
+    "//NOT-FOR-APP\n",
+    "timeoutMillisecs = Some(10000)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Data exploration"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "First, we simply iterate over the batch RDDs and print out their content, just to explore the data in our Kafka queue. We also don't need this when we deploy the notebook as application. So we tag the cell accordingly.\n",
+    "\n",
+    "**Important note**: Before you execute the cell below, you will need to start the producer container that writes into the kafka queue. When the IoT producer container is not started or has already terminated, the streaming computation in the following cell will hang and never return."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 9,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "application/vnd.jupyter.widget-view+json": {
+       "model_id": "",
+       "version_major": 2,
+       "version_minor": 0
+      },
+      "text/plain": [
+       "FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    },
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "java.lang.AbstractMethodError: org/apache/spark/internal/Logging.initializeLogIfNecessary(ZZ)Z\n",
```

Comment:
This doesn't look good :-(

Reply:
indeed. in fact this part doesn't work and I don't know why yet.